modules/database/src/ioc/rsrv/camessage.c

```diff
@@ -2434,24 +2438,77 @@ int camessage ( struct client *client )
         msg.m_cid       = ntohl ( mp->m_cid );
         msg.m_available = ntohl ( mp->m_available );
 
-        if ( CA_V49(client->minor_version_number) && msg.m_postsize == 0xffff  ) {
+        /* disconnect clients that send an invalid command */
+        if (msg.m_cmmd > CA_PROTO_LAST_CMMD) {
```

Comment:
I had deliberately avoided adding this test as a way to preserve the possibility of future extensions of the protocol.

It this point in the lifetime of CA, it's probably reasonable to expect that there won't be too many more.  So maybe this test could be adjusted to `CA_PROTO_LAST_CMMD + 16`?

Reply:
But this particular version of the server only implements `CA_PROTO_LAST_CMMD` commands. So, even if a future client, using a newer version of the protocol with more commands, if it sends a request for a command not implemented in this version (i.e. `m_cmmd > CA_PROTO_LAST_CMMD`), that request should still be rejected as the server does not implement it. Of course, command lower than `CA_PROTO_LAST_CMMD` will still be processed.

In order to use those new commands, you will need a server using the new protocol version as well. And that server will have a higher `CA_PROTO_LAST_CMMD`.

What do you think?

Reply:
We generally don't want to _disconnect_ clients or servers just because they support a newer protocol version than we do; as long as they exchange version numbers the newer end is supposed to adapt its behavior to the older end, so both can still talk using the older of the two protocol versions. That's why the RSRV code is littered with version checks. It has always been possible for different version clients and servers to talk because they are designed for exactly that.

Reply:
Oh, yes, that is a good point. Then, what about doing the same that was done for deprecated clients:
https://github.com/epics-base/epics-base/blob/ca3ef9c61e57a7413ad9626f17ce9205cdabdc5f/modules/database/src/ioc/rsrv/camessage.c#L2456-L2467

But with a different error message (`CAS: Invalid command rejected`)?

Reply:
> .. request should still be rejected as the server does not implement it.

Yes, but silently.  Otherwise introducing a new message will result in copious log spam.

This would be especially acute with UDP broadcasts.  How can I reasonably introduce a new UDP search message (eg. maybe search w/ IPv6 address) without flooding the logs of all existing IOCs?

While I know that is not the case now.  As a general rule, I think that validation errors with UDP packets should be silent by default.

Reply:
That makes sense to me. The invalid commands can then be rejected, without disconnecting the client and without generating a log error; only sending the error back to the client? 

It will look something like this:
```c
/*
 * Reject invalid commands.
 * Send the error to the client, but keep the connection open to avoid a
 * re-connect loops.
 * Do not generate a log message for this type of error to avoid flooding
 * the logs.
 */
if (msg.m_cmmd > CA_PROTO_LAST_CMMD) {
    SEND_LOCK(client);
    send_err(&msg, ECA_NOSUPPORT, client, "CAS: Invalid command rejected");
    SEND_UNLOCK(client);
    client->recvBytesToDrain = msgsize - bytes_left;
    client->recv.stk = client->recv.cnt;
    status = RSRV_OK;
    break;
}
```

Reply:
I think we need to be clear about distinguishing TCP from UDP since `camessage()` handles both.  What you show is reasonable for a TCP connection.  It isn't reasonable for UDP where responding with an error to an unknown or even invalid UDP packet would invite the sort of "amplification" DDoS scenarios seen with other UDP based protocols like DNS or NTP.  Especially if responding to a broadcast.

So I think the new validations you add are totally reasonable.  Sending a remote error to a TCP peer is also reasonable (eg. `client->proto==IPPROTO_TCP`).

Maybe these console log messages should be disabled by default?  eg. conditioned on a new `rsrvDebug` global.

Reply:
Ah, yes.  I had forgotten about that, and looked at the wrong end of the alphabet.

Reply:
I think I added all the requested changes this section. Invalid commands are now rejected without disconnecting the client, the error is send only to TCP clients, and the log message is disabled by default but can be enabled (only for TCP clients) using the `CASDEBUG` variable.